go/perf/suite/suite.go

```diff
@@ -0,0 +1,334 @@
+	"time"
+
+	"github.com/attic-labs/noms/go/chunks"
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/dataset"
+	"github.com/attic-labs/noms/go/marshal"
+	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/testify/assert"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/disk"
+	"github.com/shirou/gopsutil/host"
+	"github.com/shirou/gopsutil/mem"
+)
+
+var (
+	perfFlag        = flag.String("perf", "", "The database to write perf tests to. If this isn't specified, perf tests are skipped. If you want a dry run, use \"mem\" as a database")
+	perfRepeatFlag  = flag.Int("perf.repeat", 1, "The number of times to repeat each perf test")
+	testNamePattern = regexp.MustCompile("^Test[0-9]*([A-Z].*$)")
+)
+
+// PerfSuite is the core of the perf testing suite. See package documentation for details.
+type PerfSuite struct {
+	// T is the `testing.T` instance set when the suite is passed into `Run`.
+	T *testing.T
+
+	// W is the `io.Writer` to write test output, which only outputs if the verbose flag is set.
+	W io.Writer
+
+	// AtticLabs is the path to the attic-labs directory (e.g. /path/to/go/src/github.com/attic-labs).
+	AtticLabs string
+
+	// Database is a Noms database that tests can use for reading and writing. State is persisted across a single run of a suite.
+	Database datas.Database
+
+	// DatabaseSpec is the Noms spec of `Database` (typically a localhost URL).
+	DatabaseSpec string
+
+	tempFiles []*os.File
+	paused    time.Duration
+}
+
+type perfSuiteT interface {
+	Suite() *PerfSuite
+}
+
+type timeInfo struct {
+	elapsed, paused, total time.Duration
+}
+
+type testRun map[string]timeInfo
+
+type nopWriter struct{}
+
+func (r nopWriter) Write(p []byte) (int, error) {
+	return len(p), nil
+}
+
+// Run runs `suiteT` and writes results to dataset `datasetID` in the database given by the `-perf` command line flag.
+func Run(datasetID string, t *testing.T, suiteT perfSuiteT) {
+	if *perfFlag == "" {
+		return
+	}
+
+	assert := assert.New(t)
+
+	// Piggy-back off the go test -v flag.
+	verboseFlag := flag.Lookup("test.v")
+	assert.NotNil(verboseFlag)
+	verbose := verboseFlag.Value.(flag.Getter).Get().(bool)
+
+	suite := suiteT.Suite()
+	suite.T = t
+	if verbose {
+		suite.W = os.Stdout
+	} else {
+		suite.W = nopWriter{}
+	}
+
+	gopath := os.Getenv("GOPATH")
+	assert.True(gopath != "")
+	suite.AtticLabs = path.Join(gopath, "src", "github.com", "attic-labs")
+
+	// This is the database the perf test results are written to.
+	db, err := spec.GetDatabase(*perfFlag)
+	assert.NoError(err)
+
+	// This is the temporary database for tests to use.
+	server := datas.NewRemoteDatabaseServer(chunks.NewMemoryStore(), 0)
+	portChan := make(chan int)
+	server.Ready = func() { portChan <- server.Port() }
+	go server.Run()
+
+	port := <-portChan
+	suite.DatabaseSpec = fmt.Sprintf("http://localhost:%d", port)
+	suite.Database = datas.NewRemoteDatabase(suite.DatabaseSpec, "")
```

Comment:
as discussed -

* it can't be a straight memory store because spec doesn't know how to re-use those...
* and it could be a leveldb store, it just seems more realistic to use a HTTP store, nice to exercise the HTTP stack...
* by which argument perhaps I should be using an HTTP store backed by a leveldb store...
* but perhaps the practical difference here is whether the store ends up in-memory or on-disk...
* on my own laptop I was running out of HDD space but not memory, it'll be interesting to see how jenkins or whatever builder responds...
* come to think of it, memory does seem more reliable for timing?